### PR TITLE
Adding error codes to fail more gracefully

### PR DESCRIPTION
**Pull request by bowu**
This commit uses error codes for the cypher compiler and execution engine to fail more gracefully.

### CODE REVIEW

File: libquery/include/querying/Errors.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <system_error>
+
+#include "galois/config.h"
+
+namespace Querying {
```
Comment (ddn0) [inline libquery/include/querying/Errors.h:7]: Let's just stick to one namespace, `galois`. If you need a separate error code enum besides galois::ErrorCode, why not just create `galois::QueryErrorCode`?

File: libquery/src/OrderByOp.cpp [hunk 2]
```diff
@@ -63,27 +67,31 @@ galois::OrderByOp::Compute() {
       break;
     }
     default:
-      GALOIS_LOG_FATAL("Type {} is not supported", col.GetType());
+      GALOIS_LOG_ERROR("Type {} is not supported", col.GetType());
```
Comment (ddn0) [inline libquery/src/OrderByOp.cpp:70]: in other code, we have been using `GALOIS_LOG_DEBUG` for the extra details about errors

File: libquery/src/CypherCompiler.cpp [hunk 3]
```diff
@@ -964,11 +1084,14 @@ galois::CypherCompiler::CompileASTNode(CypherASTNode ast) {
       // match on the pattern; note that if the pattern has many children
       // it's handled as the node will not match any of these patterns
       // in this if-else chain
-      return CompileASTNode(pattern);
+      auto res = CompileASTNode(pattern);
+      if (!res) {
+        return res.error();
+      } else {
```
Comment (roshandathathri) [inline libquery/src/CypherCompiler.cpp:1090]:
```
}
return ResultSuccess();
```

**Review by roshandathathri — APPROVED** [review]
This is great!

File: libquery/src/CypherCompiler.cpp [hunk 4]
```diff
@@ -372,39 +385,74 @@ galois::CypherCompiler::CompilePatternPath(CypherASTNode ast, bool negate) {
...
         if (negate) {
           this->query_edges.back().edge_direction = galois::NO_DIRECTED_EDGE;
         } else {
           this->query_edges.back().edge_direction = galois::DIRECTED_EDGE;
         }
         if (direction == CYPHER_REL_OUTBOUND) {
-          CompileNodePatternPath(first, this->query_edges.back().caused_by);
-          CompileRelPatternPath(rel);
-          CompileNodePatternPath(second, this->query_edges.back().acted_on);
+          auto res =
```
Comment (ddn0) [inline libquery/src/CypherCompiler.cpp:408]: you can write these more compactly as:
```cpp
if (auto res = ...; !res) {
  return res.error();
}
```

File: libquery/src/CypherCompiler.cpp [hunk 5]
```diff
@@ -1067,7 +1189,11 @@ galois::CypherCompiler::Compile(const char* queryStr) {
 
   if (nerrors == 0) {
     // take ast, change it to a query graph
-    this->CompileAST(result);
+    auto res = CompileAST(result);
```
Comment (ddn0) [inline libquery/src/CypherCompiler.cpp:1192]: :)

File: libquery/src/LimitOp.cpp [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 uint32_t galois::LimitOp::num_limit_ops_ = 0;
 
-std::vector<galois::Operator*>
+galois::Result<std::vector<galois::Operator*>>
```
Comment (danielmawhirter) [inline libquery/src/LimitOp.cpp:5]: No control path returns a failure here

File: libquery/src/SkipOp.cpp [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 uint32_t galois::SkipOp::num_skip_ops_ = 0;
 
-std::vector<galois::Operator*>
+galois::Result<std::vector<galois::Operator*>>
```
Comment (danielmawhirter) [inline libquery/src/SkipOp.cpp:5]: ditto

File: libquery/src/CountFunc.cpp [hunk 8]
```diff
@@ -5,7 +5,7 @@
 #include "galois/DReducible.h"
 #include "galois/gIO.h"
 
-galois::Column
+galois::Result<galois::Column>
```
Comment (danielmawhirter) [inline libquery/src/CountFunc.cpp:8]: ditto

**Review by danielmawhirter — APPROVED** [review]
LGTM, some nits

File: libquery/include/querying/CypherCompiler.h [hunk 9]
```diff
@@ -200,6 +202,10 @@ class GALOIS_EXPORT CypherCompiler {
       CypherASTNode node_to_parse, bool is_return);
 
   //! Parses an order by node and return the metadata associated with it
+  // TODO (bwu): This function does not return
+  // Result<std::unique_ptr<CompilerOrderByMetadata>> because gcc-7 failed to convert
+  // std::unique_ptr<CompilerOrderByMetadata> to
+  // Result<std::unique_ptr<CompilerOrderByMetadata> while gcc-9 succeeded.
```
Comment (roshandathathri) [inline libquery/include/querying/CypherCompiler.h:208]: Is there a way to explicitly convert the type? @ddn0

Comment (ddn0) [reply libquery/include/querying/CypherCompiler.h:208]: Yeah:
https://github.com/KatanaGraph/katana/blob/master/libsupport/include/galois/Result.h#L25

Comment (danielmawhirter) [reply libquery/include/querying/CypherCompiler.h:208]: Adopting that in #556

Comment (ddn0) [reply libquery/include/querying/CypherCompiler.h:208]: Maybe we should put this hint in `CONTRIBUTING.md`?

Comment (bowu) [reply libquery/include/querying/CypherCompiler.h:208]: That's a good idea.

**Pull request merged**